ethcore/sync/src/light_sync/mod.rs

```diff
@@ -213,6 +214,44 @@ enum SyncState {
 	Rounds(SyncRound),
 }
 
+/// A wrapper around the SyncState that makes sure to
+/// update the giving reference to `is_idle`
+#[derive(Debug)]
+struct SyncStateWrapper {
```

Comment:
I am not really sure about the usefulness of this struct, it seems that `set` function is only called once in `set_state` of LightClient, since fields are private. 
Ok, I got it, it avoid bad use in future client inner implementation, maybe then we could move `is_idle` to it:
```rust
struct SyncStateWrapper {
  state: Mutex::new(SyncState::Idle),
  is_idle: Mutex::new(true),
}
```
and keep deref to `state`. Then we would have an easier `set` function (without handle ref).
This could be a bad idea if we need to keep is_idle mutex open (does not seems so).

Reply:
The thing is that you want to keep an open `SyncState` `Mutex` while `is_idle` still being accessible. With the suggested changes, you cannot have a lock on `SyncStateWrapper` (otherwise `is_idle` wouldn't be accessible), so you would have to expose the inner `SyncState` lock, which could be used to directly change the state, without updating `is_idle`. Does that make sense?

Reply:
Make sense, my suggestion was your second hypothesis (no mutex around SyncStateWrapper) , it does need to exposes the inner lock, on the other hand it keeps both locks together.
There is maybe less chance to mistake with current version, a mistake would be to pass any `&mut bool` to `set_state` which would be clearly intentional.